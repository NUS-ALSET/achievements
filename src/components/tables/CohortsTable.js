/**
 * @file CohortsTable container module
 * @author Theodor Shaytanov <theodor.shaytanov@gmail.com>
 * @created 22.02.18
 */

import React, { Fragment } from "react";
import PropTypes from "prop-types";

import Button from "@material-ui/core/Button";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";

import withStyles from "@material-ui/core/styles/withStyles";
import { Link } from "react-router-dom";

const styles = theme => ({
  link: {
    textDecoration: "none"
  },
  button: {
    margin: theme.spacing.unit
  }
});

class CohortsTable extends React.PureComponent {
  static propTypes = {
    onEditCohortClick: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
    cohorts: PropTypes.object.isRequired,
    currentUserId: PropTypes.string.isRequired
  };

  render() {
    const { cohorts, currentUserId, classes, onEditCohortClick } = this.props;

    return (
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Cohort name</TableCell>
            <TableCell>Instructor</TableCell>
            <TableCell>Actions</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {!Object.keys(cohorts).length ? (
            <TableRow>
              <TableCell colSpan={5}>Empty list</TableCell>
            </TableRow>
          ) : (
            Object.keys(cohorts)
              .map(id => ({ ...cohorts[id], id }))
              .map(cohort => (
                <TableRow key={cohort.id}>
                  <TableCell>{cohort.name}</TableCell>
                  <TableCell>{cohort.instructorName}</TableCell>
                  <TableCell>
                    <Link className={classes.link} to={`/cohorts/${cohort.id}`}>
                      <Button className={classes.button} variant="contained">
                        View
                      </Button>
                    </Link>
                    {cohort.owner === currentUserId && (
                      <Fragment>
                        <Button
                          className={classes.button}
                          onClick={() => onEditCohortClick(cohort)}
                          variant="contained"
                        >
                          Edit
                        </Button>
                        {/* Not implemented yet */ false && (
                          <Button
                            className={classes.button}
                            variant="contained"
                          >
                            Delete
                          </Button>
                        )}
                      </Fragment>
                    )}
                  </TableCell>
                </TableRow>
              ))
          )}
        </TableBody>
      </Table>
    );
  }
}

export default withStyles(styles)(CohortsTable);
