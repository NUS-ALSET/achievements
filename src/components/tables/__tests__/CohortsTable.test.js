/* eslint-disable no-magic-numbers */
import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import sinon from "sinon";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";
import TableCell from "@material-ui/core/TableCell";

import CohortsTable from "../CohortsTable";
import { TableBody, TableRow } from "@material-ui/core";

describe("<CohortsTable>", () => {
  let shallow;
  let mockDispatch;
  let cohorts;

  beforeEach(() => {
    mockDispatch = sinon.spy();
    shallow = createShallow({
      dive: true
    });
    cohorts = {
      cohort1: {
        id: "test",
        name: "Test"
      },
      cohort2: {
        id: "test2",
        name: "Test2"
      }
    };
  });

  it("renders without throwing with multiple cohorts", () => {
    const tree = renderer
      .create(
        <Router>
          <CohortsTable
            cohorts={cohorts}
            currentUserId="-DjgKGJ12hjh_"
            onEditCohortClick={mockDispatch}
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render cohorts list", () => {
    const wrapper = shallow(
      <CohortsTable
        cohorts={cohorts}
        currentUserId="-DjgKGJ12hjh_"
        onEditCohortClick={mockDispatch}
      />
    );

    // Expect the wrapper object to be defined
    expect(wrapper.find(TableBody)).toBeDefined();
    expect(wrapper.find(TableBody).find(TableRow)).toHaveLength(
      Object.keys(cohorts).length
    );
  });

  it("should render empty list", () => {
    const wrapper = shallow(
      <CohortsTable
        cohorts={{}}
        currentUserId="-hkjkjvhfi_dnvj"
        onEditCohortClick={mockDispatch}
      />
    );

    expect(
      wrapper
        .find(TableBody)
        .find(TableRow)
        .find(TableCell).length
    ).toEqual(1);
  });

  it("should render buttons for owner", () => {
    const cohorts = {
      cohort1: {
        id: "test",
        name: "Test",
        owner: "fdefjgwdwdfKGJ12hjh_"
      }
    };
    const wrapper = shallow(
      <CohortsTable
        cohorts={cohorts}
        currentUserId="-hkjkjvhfi_dnvj"
        onEditCohortClick={mockDispatch}
      />
    );

    // Expect the wrapper object to be defined
    expect(wrapper.find(TableBody)).toBeDefined();
    expect(
      wrapper
        .find(TableBody)
        .find(TableRow)
        .find(TableCell)
    ).toHaveLength(3);
  });

  it("renders without crashing when there are no cohorts", () => {
    const tree = renderer
      .create(
        <CohortsTable
          cohorts={{}}
          currentUserId="-hkjkjvhfi_dnvj"
          onEditCohortClick={mockDispatch}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly when there are multiple cohorts", () => {
    const tree = renderer
      .create(
        <Router>
          <CohortsTable
            cohorts={cohorts}
            currentUserId="-hkjkjvhfi_dnvj"
            onEditCohortClick={mockDispatch}
          />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
