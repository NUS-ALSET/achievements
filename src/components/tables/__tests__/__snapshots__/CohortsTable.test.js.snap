// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CohortsTable> renders correctly when there are multiple cohorts 1`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Cohort name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Instructor
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CohortsTable-link-1"
          href="/cohorts/cohort1"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 CohortsTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test2
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CohortsTable-link-1"
          href="/cohorts/cohort2"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 CohortsTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<CohortsTable> renders without crashing when there are no cohorts 1`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Cohort name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Instructor
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
        colSpan={5}
      >
        Empty list
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<CohortsTable> renders without throwing with multiple cohorts 1`] = `
<table
  className="MuiTable-root-3"
>
  <thead
    className="MuiTableHead-root-4"
  >
    <tr
      className="MuiTableRow-root-5 MuiTableRow-head-8"
    >
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Cohort name
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Instructor
      </th>
      <th
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-head-11"
        scope="col"
      >
        Actions
      </th>
    </tr>
  </thead>
  <tbody
    className="MuiTableBody-root-22"
  >
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CohortsTable-link-1"
          href="/cohorts/cohort1"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 CohortsTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
      </td>
    </tr>
    <tr
      className="MuiTableRow-root-5"
    >
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        Test2
      </td>
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      />
      <td
        aria-sort={null}
        className="MuiTableCell-root-10 MuiTableCell-body-12"
      >
        <a
          className="CohortsTable-link-1"
          href="/cohorts/cohort2"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root-49 MuiButton-root-23 MuiButton-contained-34 MuiButton-raised-37 CohortsTable-button-2"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-24"
            >
              View
            </span>
            <span
              className="MuiTouchRipple-root-52"
            />
          </button>
        </a>
      </td>
    </tr>
  </tbody>
</table>
`;
